node/impl/full/eth.go

```diff
@@ -980,14 +980,14 @@ func (a *EthModule) applyMessage(ctx context.Context, msg *types.Message, tsk ty
 		return nil, xerrors.Errorf("cannot get tipset: %w", err)
 	}
 
-	applyTsMessages := true
-	if os.Getenv("LOTUS_SKIP_APPLY_TS_MESSAGE_CALL_WITH_GAS") == "1" {
-		applyTsMessages = false
-	}
-
 	// Try calling until we find a height with no migration.
 	for {
-		res, err = a.StateManager.CallWithGas(ctx, msg, []types.ChainMsg{}, ts, applyTsMessages)
+		st, _, err := a.StateManager.TipSetState(ctx, ts)
```

Comment:
Working through the logic here - `TipSetState` will end up executing the messages on top of parent state if the requested tipset doesn't already have a height+1 not on a different fork that we can use to get the state from.

When you get down to the bottom of `tipSetState()`, having failed to find a valid existing height+1 state, you run `ExecuteTipSet`, which itself will call `HandleStateForks` which the `callInternal()` code above in this PR does so a bunch of work to avoid calling if it can tell that there's an "expensive" migration.

So we might need to port that logic down to here to handle the expensive-migration case.

But other than that, I think this results in roughly the same outcome of having all current tipset messages applied to produce a new state and then applying _only_ the requested `msg` on top of that.

I'm not sure if/how null rounds come in to play with all of this though, maybe it doesn't matter as per the discussion in #11205.

Reply:
@rvagg 

Given that `TipSetState ` already handles state forks for us, can we just not re-do it in `callInternal` for our use case ? We can control that behaviour in `callInternal` with a flag.

Reply:
> already handles state forks for us,

but that's the problem, imagine if we did this across an expensive migration boundary and triggered a migration just to get the state at this tipset.

All of this logic exists to check that condition: https://github.com/filecoin-project/lotus/blob/187e837ab8bf8261bc704daeb42de7348c294274/chain/stmgr/call.go#L104-L132

But that's only applied after you've potentially executed that tipset because `TipSetState` can result in an `ExecuteTipSet` if it doesn't already have it handy, that executes `ApplyBlocks` which in turn runs `HandleStateForks`.

So I **_think_**, if you run this, it'll still do what it says in the comment at the top of the `for` ("Try calling until we find a height with no migration."), but it'll actually perform a migration _before_ it checks whether there's an expensive migration.

I believe what we want is to run an `hasExpensiveForkBetween(pts.Height(), ts.Height()+1)` ourselves and walk back up parents until it's `false`. But currently `hasExpensiveForkBetween` is a private method of `StateManager`.

Reply:
Ahaa good catch @rvagg I have fixed this.